### PR TITLE
Make giving null to thrown() consistent with mock object creation

### PR DESCRIPTION
**Pull request by Vampire**

### CODE REVIEW

**Comment by Vampire** [comment]
Current dependencies on/for this PR:
* master
  * **PR #1784** <a href="https://app.graphite.dev/github/pr/spockframework/spock/1784" target="_blank">Graphite</a> 
    * **PR #1799** <a href="https://app.graphite.dev/github/pr/spockframework/spock/1799" target="_blank">Graphite</a>  👈
    * **PR #1785** <a href="https://app.graphite.dev/github/pr/spockframework/spock/1785" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/spockframework/spock/1799?utm_source=stack-comment).

File: spock-specs/src/test/groovy/org/spockframework/smoke/condition/ExceptionConditions.groovy [hunk 1]
```diff
@@ -168,6 +168,59 @@ thrown() == e
...
+  }
+
+  def "thrown condition with only explicit type null"() {
+    when:
+    runner.runFeatureBody """
+when:
+true
+
+then:
+thrown(null)
```
Comment (szpak) [inline spock-specs/src/test/groovy/org/spockframework/smoke/condition/ExceptionConditions.groovy:193]: Do you think that test case for `def e = thrown(null)` would be useful?

File: spock-specs/src/test/groovy/org/spockframework/smoke/condition/ExceptionConditions.groovy [hunk 2]
```diff
@@ -168,6 +168,59 @@ thrown() == e
...
+    """
+
+    then:
+    InvalidSpecException e = thrown()
+    e.message == "Thrown exception type cannot be inferred automatically. Please specify a type explicitly (e.g. 'thrown(MyException)')."
```
Comment (szpak) [inline spock-specs/src/test/groovy/org/spockframework/smoke/condition/ExceptionConditions.groovy:198]: Those 2 cases are similar. Maybe `where:` could be used to reduce duplication (unless, you want to emphasis by the name of the test what case is exactly covered here).

**Review by szpak — DISMISSED** [review]

File: spock-core/src/main/java/org/spockframework/lang/SpecInternals.java [hunk 3]
```diff
@@ -64,26 +64,25 @@ <T> T oldImpl(T expression) {
...
-      throw new InvalidSpecException("Thrown exception type cannot be inferred automatically. " +
-        "Please specify a type explicitly (e.g. 'thrown(MyException)').");
-    }
-
     return inferredType.cast(checkExceptionThrown(inferredType));
   }
 
   Throwable thrownImpl(String inferredName, Class<? extends Throwable> inferredType, Class<? extends Throwable> specifiedType) {
-    return checkExceptionThrown(specifiedType);
+    return checkExceptionThrown(specifiedType != null ? specifiedType : inferredType);
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/lang/SpecInternals.java:71]:
```suggestion
    return checkExceptionThrown(specifiedType == null ? inferredType : specifiedType);
```

**Review by leonard84 — DISMISSED** [review]

**Review by leonard84 — APPROVED** [review]

File: spock-core/src/main/java/org/spockframework/lang/SpecInternals.java (hunk 3 above)
Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/lang/SpecInternals.java]: Just wanted to be consistent with `createMockImpl`, but there it is of course a bit different as there are three alternatives.

**Pull request merged**